### ISSUE TITLE
Unable to use MPI.jl on more than one process on a cluster: segmentation fault

### ISSUE THREAD

**Issue by viaudg** [JuliaParallel/MPI.jl#171]
I compiled julia-0.5.0 on a computing cluster with `intel-mkl/11.3.3`, `intel-compilers/16.0.3`, `python/2.7.12`, `cmake/3.5.0`. It worked up to the exception that during the `make install` I got this error for `julia-debug`:

```
/bin/sh: line 1: 26678 Segmentation fault      (core dumped) /gpfs/workdir/viaudg/build/julia-0.5.0/usr/bin/julia-debug -O0 -C native --output-o /gpfs/workdir/viaudg/build/julia-0.5.0/usr/lib/julia/sys-debug.o --startup-file=no --sysimage /gpfs/workdir/viaudg/build/julia-0.5.0/usr/lib/julia/inference.ji sysimg.jl
*** This error is usually fixed by running `make clean`. If the error persists, try `make cleanall`. ***
make[2]: *** [/gpfs/workdir/viaudg/build/julia-0.5.0/usr/lib/julia/sys-debug.o] Error 1
make[1]: *** [julia-sysimg-debug] Error 2
make: *** [install] Error 2
```

which did not prevent me to have a working version of `julia` available from the build directory `/gpfs/workdir/viaudg/build/julia-0.5.0/usr/bin/`..!

I must mention that I managed to run an MPI code with julia-0.4.6 on the same cluster without problem. So then I added the latest version of the MPI package for julia-0.5.0 successfully. The module I use on this cluster for MPI is `sgi-mpt/2.12`. I tried to run the example `mpi-hello.jl` and several cases arise:
- when I launch it directly on the node with `mpirun -n 1 julia mpi-hello.jl`, it works 100% of the time
- when I launch it directly on the node with `mpirun -n 2 julia mpi-hello.jl`, I have the correct output `"Hello world, I am 1 of 2Hello world, I am 0 of 2"` but I still get this error at the end:

```
julia: /gpfs/workdir/viaudg/build/julia-0.5.0/deps/srccache/libuv-8d5131b6c1595920dd30644cd1435b4f344b46c8/src/unix/linux-core.c:329: uv__io_poll: Assertion `(unsigned) fd < loop->nwatchers' failed.
MPT: On host fusion, Program /gpfs/workdir/viaudg/build/julia-0.5.0/usr/bin/julia, Rank 1, Process 28399 received signal SIGABRT/SIGIOT(6)
MPT: --------stack traceback-------
MPT: Attaching to program: /proc/28399/exe, process 28399
MPT: [Thread debugging using libthread_db enabled]
MPT: done.
MPT: Try: yum --enablerepo='*-debug*' install /usr/lib/debug/.build-id/46/1f2bae0a837a77e768ee843e127853b266e78e.debug
MPT: (no debugging symbols found)...done.
MPT: Try: yum --enablerepo='*-debug*' install /usr/lib/debug/.build-id/0e/c1d06f20ed0d15b5380d27566a0847699480bb.debug
MPT: (no debugging symbols found)...done.
MPT: 0x00007fffecc372ad in waitpid () from /lib64/libpthread.so.0
MPT: warning: File "/gpfs/opt/compilers/gcc/5.4.0/lib64/libstdc++.so.6.0.21-gdb.py" auto-loading has been declined by your `auto-load safe-path' set to "/usr/share/gdb/auto-load:/usr/lib/debug:/usr/bin/mono-gdb.py".
MPT: To enable execution of this file add
MPT:    add-auto-load-safe-path /gpfs/opt/compilers/gcc/5.4.0/lib64/libstdc++.so.6.0.21-gdb.py
MPT: line to your configuration file "/home/viaudg/.gdbinit".
MPT: To completely disable this security protection add
MPT:    set auto-load safe-path /
MPT: line to your configuration file "/home/viaudg/.gdbinit".
MPT: For more information about this security protection see the
MPT: "Auto-loading safe path" section in the GDB manual.  E.g., run from the shell:
MPT:    info "(gdb)Auto-loading safe path"
MPT: Missing separate debuginfos, use: debuginfo-install glibc-2.12-1.149.el6.x86_64 keyutils-libs-1.4-5.el6.x86_64 krb5-libs-1.10.3-33.el6.x86_64 libbitmask-2.0-sgi712r2.rhel6.x86_64 libcom_err-1.41.12-21.el6.x86_64 libcpuset-1.0-sgi712r4.rhel6.x86_64 libibverbs-1.1.8-3.el6.x86_64 libmlx4-1.0.6-6.el6.x86_64 libmlx5-1.0.1-1.el6.x86_64 libmthca-1.0.6-3.el6.x86_64 libnl-1.1.4-2.el6.x86_64 libselinux-2.0.94-5.8.el6.x86_64 xpmem-1.6-sgi712r6.rhel6.x86_64 zlib-1.2.3-29.el6.x86_64
MPT: (gdb) #0  0x00007fffecc372ad in waitpid () from /lib64/libpthread.so.0
MPT: #1  0x00007ffdc02c26bc in mpi_sgi_system (header=<value optimized out>) at sig.c:99
MPT: #2  MPI_SGI_stacktraceback (header=<value optimized out>) at sig.c:319
MPT: #3  0x00007ffdc02c31fe in first_arriver_handler (signo=6, stack_trace_sem=0x7ffdbede0500) at sig.c:468
MPT: #4  0x00007ffdc02c3452 in slave_sig_handler (signo=6, siginfo=<value optimized out>, extra=<value optimized out>) at sig.c:542
MPT: #5  <signal handler called>
MPT: #6  0x00007fffec322625 in raise () from /lib64/libc.so.6
MPT: #7  0x00007fffec323e05 in abort () from /lib64/libc.so.6
MPT: #8  0x00007fffec31b74e in __assert_fail_base () from /lib64/libc.so.6
MPT: #9  0x00007fffec31b810 in __assert_fail () from /lib64/libc.so.6
MPT: #10 0x00007fffed3de7f1 in uv__io_poll (loop=0x7fffed8d5c60, timeout=0) at /gpfs/workdir/viaudg/build/julia-0.5.0/deps/srccache/libuv-8d5131b6c1595920dd30644cd1435b4f344b46c8/src/unix/linux-core.c:329
MPT: #11 0x00007fffed3c7a9e in uv_run (loop=0x7fffed8d5c60, mode=UV_RUN_ONCE) at /gpfs/workdir/viaudg/build/julia-0.5.0/deps/srccache/libuv-8d5131b6c1595920dd30644cd1435b4f344b46c8/src/unix/core.c:354
MPT: #12 0x00007fffce26ec41 in julia_process_events_20256 () from /gpfs/workdir/viaudg/build/julia-0.5.0/usr/lib/julia/sys.so
MPT: #13 0x00007fffce26f0ed in julia_wait_20253 () from /gpfs/workdir/viaudg/build/julia-0.5.0/usr/lib/julia/sys.so
MPT: #14 0x00007fffce26f959 in julia_wait_20251 () from /gpfs/workdir/viaudg/build/julia-0.5.0/usr/lib/julia/sys.so
MPT: #15 0x00007fffce26fa0c in jlcall_wait_20251 () from /gpfs/workdir/viaudg/build/julia-0.5.0/usr/lib/julia/sys.so
MPT: #16 0x00007fffed2968be in jl_call_method_internal (args=0x6eef, nargs=28399) at /gpfs/workdir/viaudg/build/julia-0.5.0/src/julia_internal.h:189
MPT: #17 jl_apply_generic (args=0x6eef, nargs=28399) at /gpfs/workdir/viaudg/build/julia-0.5.0/src/gf.c:1942
MPT: #18 0x00007ffdc0a6e98a in julia_#592_70966 ()
MPT: #19 0x00007ffdc0a6e9a0 in jlcall_#592_70966 ()
MPT: #20 0x00007fffed2968be in jl_call_method_internal (args=0x6eef, nargs=28399) at /gpfs/workdir/viaudg/build/julia-0.5.0/src/julia_internal.h:189
MPT: #21 jl_apply_generic (args=0x6eef, nargs=28399) at /gpfs/workdir/viaudg/build/julia-0.5.0/src/gf.c:1942
MPT: #22 0x00007fffed2bcade in start_task () at /gpfs/workdir/viaudg/build/julia-0.5.0/src/task.c:253
MPT: #23 0x0000000000000000 in ?? ()
MPT: (gdb) A debugging session is active.
MPT: 
MPT:    Inferior 1 [process 28399] will be detached.
MPT: 
MPT: Quit anyway? (y or n) [answered Y; input not from terminal]
MPT: Detaching from program: /proc/28399/exe, process 28399
MPT: -----stack traceback ends-----
MPT: On host fusion, Program /gpfs/workdir/viaudg/build/julia-0.5.0/usr/bin/julia, Rank 1, Process 28399: Dumping core on signal SIGABRT/SIGIOT(6) into directory /gpfs/home/viaudg/laurent
signal (6): Aborted
while loading /gpfs/home/viaudg/laurent/mpi-hello.jl, in expression starting on line 16
gsignal at /lib64/libc.so.6 (unknown line)
abort at /lib64/libc.so.6 (unknown line)
unknown function (ip: 0x7fffec31b74d)
__assert_fail at /lib64/libc.so.6 (unknown line)
uv__io_poll at /gpfs/workdir/viaudg/build/julia-0.5.0/deps/srccache/libuv-8d5131b6c1595920dd30644cd1435b4f344b46c8/src/unix/linux-core.c:329
uv_run at /gpfs/workdir/viaudg/build/julia-0.5.0/deps/srccache/libuv-8d5131b6c1595920dd30644cd1435b4f344b46c8/src/unix/core.c:354
process_events at ./libuv.jl:82
wait at ./event.jl:147
wait at ./event.jl:27
unknown function (ip: 0x7fffce26fa0b)
jl_call_method_internal at /gpfs/workdir/viaudg/build/julia-0.5.0/src/julia_internal.h:189 [inlined]
jl_apply_generic at /gpfs/workdir/viaudg/build/julia-0.5.0/src/gf.c:1942
macro expansion at ./multi.jl:887 [inlined]
#592 at ./event.jl:68
unknown function (ip: 0x7ffdc0a6e99f)
jl_call_method_internal at /gpfs/workdir/viaudg/build/julia-0.5.0/src/julia_internal.h:189 [inlined]
jl_apply_generic at /gpfs/workdir/viaudg/build/julia-0.5.0/src/gf.c:1942
start_task at /gpfs/workdir/viaudg/build/julia-0.5.0/src/task.c:253
unknown function (ip: 0xffffffffffffffff)
Allocations: 1036740 (Pool: 1035886; Big: 854); GC: 0
MPT: MPI_COMM_WORLD rank 1 has terminated without calling MPI_Finalize()
    aborting job
MPT: Received signal 6
```
- when I submit it via PBSPro with `qsub mpi-hello.pbs` and this file contains `mpirun -n 1 julia mpi-hello.jl`, it works every time,
- when I submit it via PBSPro with `qsub mpi-hello.pbs` and this file contains `mpirun -n 2 julia mpi-hello.jl`, I get this error:

```
signal (11): Segmentation fault
while loading /gpfs/home/viaudg/laurent/mpi-hello.jl, in expression starting on line 2
unknown function (ip: 0xffffffffffffffff)
jl_call_method_internal at /gpfs/workdir/viaudg/build/julia-0.5.0/src/julia_internal.h:189 [inlined]
jl_apply_generic at /gpfs/workdir/viaudg/build/julia-0.5.0/src/gf.c:1942
do_call at /gpfs/workdir/viaudg/build/julia-0.5.0/src/interpreter.c:66
eval at /gpfs/workdir/viaudg/build/julia-0.5.0/src/interpreter.c:190
eval_body at /gpfs/workdir/viaudg/build/julia-0.5.0/src/interpreter.c:469
jl_interpret_call at /gpfs/workdir/viaudg/build/julia-0.5.0/src/interpreter.c:573 [inlined]
jl_interpret_toplevel_thunk at /gpfs/workdir/viaudg/build/julia-0.5.0/src/interpreter.c:580
jl_toplevel_eval_flex at /gpfs/workdir/viaudg/build/julia-0.5.0/src/toplevel.c:572 [inlined]
jl_toplevel_eval at /gpfs/workdir/viaudg/build/julia-0.5.0/src/toplevel.c:580
jl_toplevel_eval_in_warn at /gpfs/workdir/viaudg/build/julia-0.5.0/src/builtins.c:590
__init__ at /workdir/viaudg/opt/julia-pkg/v0.5/MPI/src/MPI.jl:34
unknown function (ip: 0x2aacc5845b8f)
jl_call_method_internal at /gpfs/workdir/viaudg/build/julia-0.5.0/src/julia_internal.h:189 [inlined]
jl_apply_generic at /gpfs/workdir/viaudg/build/julia-0.5.0/src/gf.c:1942
jl_module_run_initializer at /gpfs/workdir/viaudg/build/julia-0.5.0/src/toplevel.c:83
jl_init_restored_modules at /gpfs/workdir/viaudg/build/julia-0.5.0/src/dump.c:1994 [inlined]
_jl_restore_incremental at /gpfs/workdir/viaudg/build/julia-0.5.0/src/dump.c:2560 [inlined]
jl_restore_incremental at /gpfs/workdir/viaudg/build/julia-0.5.0/src/dump.c:2580
_include_from_serialized at ./loading.jl:150
_require_from_serialized at ./loading.jl:187
_require_search_from_serialized at ./loading.jl:217
unknown function (ip: 0x2aaab67798d0)
jl_call_method_internal at /gpfs/workdir/viaudg/build/julia-0.5.0/src/julia_internal.h:189 [inlined]
jl_apply_generic at /gpfs/workdir/viaudg/build/julia-0.5.0/src/gf.c:1942
require at ./loading.jl:371
unknown function (ip: 0x2aaab655d8cb)
jl_call_method_internal at /gpfs/workdir/viaudg/build/julia-0.5.0/src/julia_internal.h:189 [inlined]
jl_apply_generic at /gpfs/workdir/viaudg/build/julia-0.5.0/src/gf.c:1942
jl_apply at /gpfs/workdir/viaudg/build/julia-0.5.0/src/julia.h:1392 [inlined]
eval_import_path_ at /gpfs/workdir/viaudg/build/julia-0.5.0/src/toplevel.c:402 [inlined]
eval_import_path at /gpfs/workdir/viaudg/build/julia-0.5.0/src/toplevel.c:429 [inlined]
jl_toplevel_eval_flex at /gpfs/workdir/viaudg/build/julia-0.5.0/src/toplevel.c:480
jl_parse_eval_all at /gpfs/workdir/viaudg/build/julia-0.5.0/src/ast.c:717
jl_load at /gpfs/workdir/viaudg/build/julia-0.5.0/src/toplevel.c:596
jl_load_ at /gpfs/workdir/viaudg/build/julia-0.5.0/src/toplevel.c:605
include_from_node1 at ./loading.jl:488
unknown function (ip: 0x2aaab652284b)
jl_call_method_internal at /gpfs/workdir/viaudg/build/julia-0.5.0/src/julia_internal.h:189 [inlined]
jl_apply_generic at /gpfs/workdir/viaudg/build/julia-0.5.0/src/gf.c:1942
process_options at ./client.jl:262
_start at ./client.jl:318
unknown function (ip: 0x2aaab6548068)
jl_call_method_internal at /gpfs/workdir/viaudg/build/julia-0.5.0/src/julia_internal.h:189 [inlined]
jl_apply_generic at /gpfs/workdir/viaudg/build/julia-0.5.0/src/gf.c:1942
unknown function (ip: 0x401e03)
unknown function (ip: 0x401d6d)
__libc_start_main at /lib64/libc.so.6 (unknown line)
unknown function (ip: 0x401b68)
Allocations: 1023383 (Pool: 1022529; Big: 854); GC: 0
```

So basically, I'm not able to use several processes with MPI in Julia. In the last case, which is the one that annoys me the most, it seems that the problem lies at `src/MPI.jl` line 34 when looking up the dictionary `_mpi_functions`:

```
for (jname, fname) in _mpi_functions
    eval(:(const $jname = Libdl.dlsym(libmpi_handle, $fname)))
end
```

I have no idea where the problem comes from. If someone has an idea, I'd be very thankful.

**Comment by andreasnoack** [comment]
Could you show us the content of `mpi-hello.jl`. It might also be worth trying out the generic binaries instead of building from source.

**Comment by viaudg** [comment]
The content of `mpi-hello.jl` was the same as in the examples:
```
using MPI

function do_hello()
    comm = MPI.COMM_WORLD
    println("Hello world, I am $(MPI.Comm_rank(comm)) of $(MPI.Comm_size(comm))")
    MPI.Barrier(comm)
end

function main()
    MPI.Init()
    do_hello()
    MPI.Finalize()
end

main()
```
but just to be sure I tried with a simple `using MPI` in the file, so the problem really is the module loading.

I tried by downloading the binary version of `julia-0.5.0` here `https://julialang.s3.amazonaws.com/bin/linux/x64/0.5/julia-0.5.0-linux-x86_64.tar.gz` and tried the same four test cases, and the outputs are exactly the same in all cases.

As far as I'm concerned I went back to `julia-0.4.6` for my current work so resolving this issue is not crucial to me, but I thought it would be interesting to find out why this does not work.

**Comment by viaudg** [comment]
I narrowed down the sources of the problem: everything works for both versions of `julia-0.5.0` (compiled from source and binary) if I use the MPI version of Intel `intel-mpi/5.1.2` instead of that of SGI `sgi-mpt/2.12`.

**Comment by michel2323** [comment]
I can confirm this on a Cray XC40 with the gnu environment with julia 0.6.4. This does not happen with julia 0.4.7, although the differences in the MPI.jl code is marginal.

**Comment by barche** [comment]
Could it be related to [this bug](http://trac.mpich.org/projects/mpich/ticket/2196) that was fixed in mpich 3.2? This gave segfaults on older versions of mpich on other platforms. I don't have a Cray handy, so I don't know how closely their mpich follows the original.

**Comment by michel2323** [comment]
I doubt it, since it works with Julia 0.4.7 and the same system environment. But who knows, I have no other choice than running cray-mpich right now.

It crashes in MPI.jl:53 , but at times also somewhere else. And it only happens with more than 1 process.

**Comment by michel2323** [comment]
I tested now all available Cray MPI version, while going back to commit 6efa4902c4c03a6984ce4d29e3370ff1b11665ee of MPI.jl . Bottom line is that with the same MPI.jl and the same MPI libraries, Julia 0.4.7 runs and Julia 0.6.x crashes in that loop at MPI.jl:34 .

Is there anything I could try as a workaround? I tried obviously commenting out that loop, but then it crashes later when an actual symbol is resolved.

**Comment by simonbyrne** [comment]
It's unclear exactly what the issue was here. #271 changed the internals quite a bit, so if there are still problems, please open a new issue.
